source/content/tools/pytest-api-examples.mkd

```diff
@@ -0,0 +1,321 @@
+[pytest fixture]: https://docs.pytest.org/en/latest/reference.html#pytest-fixture
+
+
+~~~ python
+@pytest.fixture
+def example_data_file_with_teardown() -> typing.Generator[pathlib.Path, None, None]:
+    """Yield a large file, then delete it after each test completes."""
+    large_data_file = fetch_file_from_s3()
+    yield large_data_file
+    print("Cleaning up file: {}".format(large_data_file))
+    large_data_file.unlink()
+
+
+def test_fixture_teardown_1(example_data_file_with_teardown: pathlib.Path):
+    assert example_data_file_with_teardown.exists()
+
+
+def test_fixture_teardown_2(example_data_file_with_teardown: pathlib.Path):
+    assert example_data_file_with_teardown.exists()
+~~~
+
+    Fetching a large file from S3.
+    Cleaning up file: /var/folders/kw/hfflzwfs3xl1_xzq83270dh40000gn/T/tmpfsgb870s
+    Fetching a large file from S3.
+    Cleaning up file: /var/folders/kw/hfflzwfs3xl1_xzq83270dh40000gn/T/tmp4_k7d8wn
+
+
+
+## Scoping fixtures
+
+Documentation: [scope sharing][]
+
+[scope sharing]: https://docs.pytest.org/en/latest/fixture.html#scope-sharing-fixtures-across-classes-modules-packages-or-session
+
+In the example above the code after the `yield` runs every time the fixture is
+used, this might be inappropriate if the fixture computationally expensive. An
+alternative to caching the result (described below), would be to set the scope
+of the fixture with `pytest.fixture(scope=...)`. For example
+`pytest.fixture(scope="session")` will run only once for the entire pytest
+session. Possible values for `scope=...` are
+`["class", "module", "package", "session"]`. A `Callable` can also be passed
+which will be evaluated once, see [dynamic scope][].
+
+[dynamic scope]: https://docs.pytest.org/en/latest/fixture.html#dynamic-scope
+
+
+~~~ python
+@pytest.fixture(scope="session")
+def example_data_file_for_session() -> typing.Generator[pathlib.Path, None, None]:
+    """Yield a large file, then delete it after the test completes."""
+    large_data_file = fetch_file_from_s3()
+    yield large_data_file
+    print("Cleaning up file: {}".format(large_data_file))
+    large_data_file.unlink()
+
+
+def test_fixture__session_teardown_1(example_data_file_for_session: pathlib.Path):
+    print("Running test 1")
+    assert example_data_file_for_session.exists()
+
+
+def test_fixture_session_teardown_2(example_data_file_for_session: pathlib.Path):
+    print("Running test 2")
+    assert example_data_file_for_session.exists()
+~~~
+
+    Fetching a large file from S3.
+    Running test 1
+    Running test 2
+    Cleaning up file: /var/folders/kw/hfflzwfs3xl1_xzq83270dh40000gn/T/tmpmdzgjvda
+
+
+
+## Parameterising fixtures
+
+If you find yourself using the same `pytest.mark.parametrize` arguments multiple
+times in your tests, this can be refactored into a fixture using
+`pytest.fixture(params=...)`
+
+
+~~~ python
+@pytest.fixture(params=["", tempfile.mkdtemp(), "/non_existing_file"])
```

Comment:
`mkdtemp` feels like it should be in a context in a fixture to make sure it gets cleaned up afterwards